aten/src/ATen/core/dispatch/KernelFunction.h

```diff
@@ -2,15 +2,207 @@
+  static KernelFunction makeFromUnboxedOnlyFunctor(std::shared_ptr<KernelFunctor> kernelFunctor) {
+    // TODO We want to get rid of kernels that have only an unboxed function pointer.
+    //      All kernels should have a boxed pointer.
+
+    static_assert(guts::is_functor<KernelFunctor>::value, "Tried to call KernelFunction::makeFromUnboxedFunctor<KernelFunctor> but the argument is not a functor.");
+    static_assert(std::is_base_of<OperatorKernel, KernelFunctor>::value, "Tried to call KernelFunction::makeFromUnboxedFunctor<KernelFunctor>, but the functor doesn't inherit from c10::OperatorKernel. Please have the functor inherit from it.");
+
+    return KernelFunction(
+      std::move(kernelFunctor),
+      nullptr, // Don't create a boxed kernel for this
+      reinterpret_cast<void*>(&detail::wrap_kernel_functor_unboxed<KernelFunctor>::call),
+      detail::hashFunctionSignature<KernelFunctor>()
+    );
+  }
+
+  template<class FuncType, FuncType* func, bool AllowLegacyTypes = false>
+  static KernelFunction makeFromUnboxedFunction() {
+    static_assert(guts::is_function_type<FuncType>::value, ""); // TODO
+    static_assert(!std::is_same<FuncType, BoxedKernelFunction>::value, ""); // TODO
+    static_assert(func != nullptr, "Kernel function cannot be nullptr");
+
+    return makeFromUnboxedFunctor<AllowLegacyTypes>(
+      std::make_shared<typename detail::WrapKernelFunction<FuncType, func>::type>()
+    );
+  }
+
+  template<class FuncType, FuncType* func>
+  static KernelFunction makeFromUnboxedOnlyFunction() {
+    // TODO We want to get rid of kernels that have only an unboxed function pointer.
+    //      All kernels should have a boxed pointer.
+
+    static_assert(guts::is_function_type<FuncType>::value, ""); // TODO
+    static_assert(!std::is_same<FuncType, BoxedKernelFunction>::value, ""); // TODO
+    static_assert(func != nullptr, "Kernel function cannot be nullptr");
+
+    return makeFromUnboxedOnlyFunctor(
+      std::make_shared<typename detail::WrapKernelFunction<FuncType, func>::type>()
+    );
+  }
+
+  template<bool AllowLegacyTypes = false, class FuncType>
+  static KernelFunction makeFromUnboxedRuntimeFunction(FuncType* func) {
+    static_assert(guts::is_function_type<FuncType>::value, ""); // TODO
+    static_assert(!std::is_same<FuncType, BoxedKernelFunction>::value, ""); // TODO
+    TORCH_INTERNAL_ASSERT(func != nullptr, "Kernel function cannot be nullptr");
+
+    return makeFromUnboxedFunctor<AllowLegacyTypes>(
+      std::make_shared<detail::WrapRuntimeKernelFunctor<guts::decay_t<FuncType>>>(func)
+    );
+  }
+
+  template<bool AllowLegacyTypes = false, class Lambda>
+  static KernelFunction makeFromUnboxedLambda(Lambda&& lambda) {
+    static_assert(guts::is_functor<guts::decay_t<Lambda>>::value, ""); // TODO
+
+    return makeFromUnboxedFunctor<AllowLegacyTypes>(
+      std::make_shared<detail::WrapRuntimeKernelFunctor<guts::decay_t<Lambda>>>(std::forward<Lambda>(lambda))
+    );
+  }
```

Comment:
Is it possible to reduce the surface area of this API? It was very irritating to have to fix multiple dispatch tests three times, because there were three copy pasted versions of morally the same test.

Reply:
I don't think so unfortunately. The paths for function schema inference, for example, are slightly different for lambdas vs functions, and while writing that code, it did happen to me from time to time that only the test case for one of the APIs failed because I had forgotten something. I feel safer with all APIs being fully tested.